### PR TITLE
JDK-8300080: offset_of for GCC/Clang exhibits undefined behavior and is not always a compile-time constant

### PR DESCRIPTION
**Pull request by jcking**
The implementation of `offset_of` for GCC/Clang only deals with types are aligned to 16 bytes or less, if they are more, such as `zCollectedHeap` the behavior is undefined. UBSan also suggests that `offset_of` is not always a compile time constant, as the stack trace came from the dynamic loader during library loading. This patch changes `offset_of` to use `offsetof` and disables the warning `invalid-offsetof` for the JVM.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8300080](https://bugs.openjdk.org/browse/JDK-8300080): offset_of for GCC/Clang exhibits undefined behavior and is not always a compile-time constant

### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/11978/head:pull/11978` \
`$ git checkout pull/11978`

Update a local copy of the PR: \
`$ git checkout pull/11978` \
`$ git pull https://git.openjdk.org/jdk.git pull/11978/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11978`

View PR using the GUI difftool: \
`$ git pr show -t 11978`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11978.diff">https://git.openjdk.org/jdk/pull/11978.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/11978#issuecomment-1380970396)

### CODE REVIEW

**Comment by jcking** [comment]
@aph FYI undoes `offset_of` change from JDK-8294902

**Comment by shipilev** [comment]
You need to page @theRealAph for this.

**Comment by kimbarrett** [comment]
Please don't make this change.  We are intentionally not using offsetof.  See JDK-8239357.

A change to the existing code would be to use `alignof(klass)` to align the space buffer, once use of
`alignof` is approved.  That would fix the alignment problem.

**Comment by kimbarrett** [comment]
Also, the FIXME comment added by JDK-8294902 suggesting using `offsetof` should be removed.

**Comment by kimbarrett** [comment]
Also @jcking , please always assign a subcompent for hotspot bugs.  I moved this one to the runtime subcomponent.

**Comment by jcking** [comment]
I read JDK-8239357 and related information. I am not sure I understand why `offsetof` is being avoided given that C++17 shifted the behavior to be `conditionally-supported`. `conditionally-supported` is a nice way of saying it's up to the compiler on whether it gives meaningful answers or fails to compile. All compilers I am aware of give meaningful answers. Were going to have to agree to disagree on the rationale of not using `offsetof`, but it appears Clang/GCC optimize it nicely at least in isolation.

Looking at godbolt, the current definition of `offset_of` produces some horrific assembly without optimizations. Thankfully with `-O2` it boils down to a constant. I am guessing UBSan prevents the optimization so it can instrument it, or rather it instruments it making it unable to optimize it. Perhaps we should apply `ALWAYSINLINE` to the lambda to get better results regardless of build mode and use `alignas` now due to `std::aligned_storage_t` being deprecated in C++23?

**Comment by TheShermanTanker** [comment]
Small nit, I think it might be better to use `ALWAYSINLINE` instead of the attribute for the custom `offset_of`, it looks somewhat neater

> Also, the FIXME comment added by JDK-8294902 suggesting using `offsetof` should be removed.

I partially agree with Justin on this one when it comes to `offsetof`, if I'm missing something important perhaps we could leave the comment as is or reword it to not be a `FIXME` while we stay on C++14, to leave it open for if HotSpot properly moves to a newer C++ version in the future. Just my 2¢ on the issue

> A change to the existing code would be to use `alignof(klass)` to align the space buffer, once use of
> `alignof` is approved. That would fix the alignment problem.

The `alignof` proposal is still awaiting approval, given that we're speaking of it :P

**Comment by jcking** [comment]
> Small nit, I think it might be better to use `ALWAYSINLINE` instead of the attribute for the custom `offset_of`, it looks somewhat neater

Unfortunately `ALWAYSINLINE` includes the `inline` keyword, which I do not believe can be used at the location we need to place the attribute.

**Comment by TheShermanTanker** [comment]
Ah, right, I completely forgot about that. I wonder why we still need the `inline` keyword though? I know for early versions of gcc you needed the `inline` declaration for always_inline to have any effect, but somewhere down the line gcc's maintainers (rightfully) decided that requiring the keyword and the attribute together was a little ridiculous, so today gcc will forcefully inline such declarations as we want it to, even with only the always_inline attribute requested without the corresponding `inline` keyword specified

File: src/hotspot/share/utilities/globalDefinitions_gcc.hpp
Comment (kimbarrett) [inline src/hotspot/share/utilities/globalDefinitions_gcc.hpp:142]: suggest s/Technically using/Using/

File: src/hotspot/share/utilities/globalDefinitions_gcc.hpp
Comment (kimbarrett) [inline src/hotspot/share/utilities/globalDefinitions_gcc.hpp:152]: The xlc version of this macro is the same as the old (before JDK-8294902) one from here, with a
comment saying it is the same as the gcc version.  Clearly that's no longer true.  Either that one
should also be updated or the comment updated.  That's probably something for the aix-ppc port
maintainers to decide.  There ought to be a followup bug report for it though.

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Comment by GoeLin** [comment]
@kimbarret, is this ready to be sponsored?

**Comment by kimbarrett** [comment]
> @kimbarret, is this ready to be sponsored?

It needs a second reviewer.

**Comment by TheShermanTanker** [comment]
Should probably change the integration requirements to mark this as such

**Comment by TheShermanTanker** [comment]
:(

**Comment by dholmes-ora** [comment]
@TheShermanTanker Hotspot changes _always_ need two reviewers, unless they meet the trivial criteria. We don't generally issue the /reviewers command.

**Pull request closed without merging**
